README.md

```diff
@@ -1 +1,569 @@
+This article was originally a workshop at QCon Shanghai 2018. Author gives us his advice for best practices writing Go code.
+
+* Think you understand the Single Responsibility Principle? https://hackernoon.com/you-dont-understand-the-single-responsibility-principle-abfdd005b137
+
+* Layered Architecture (and how it shows up in Microservice Architectures): https://medium.com/code-smells/layered-architecture-f11bc04c5d6c
+
+All of them are quite important for us as for backend developers.
+
+# Golang Coding Guidelines
+
+
+Based on following articles:
+
+* Effective Go - https://golang.org/doc/effective_go.html#interface-names
+
+* Code review comments - https://github.com/golang/go/wiki/CodeReviewComments
+
+## Avoid global variables
+
+Global variables make testing and readability hard and every method has access to them (even those, that don't need it).
+Use structs to encapsulate the variables and make them available only to those functions that actually need them by making them methods implemented for that struct.
+
+Alternatively higher-order functions can be used to inject dependencies via closures.
+
+Bad:
+
+```Go
+var db *sql.DB
+
+func main() {
+	db = // ...
+	http.HandleFunc("/drop", DropHandler)
+	// ...
+}
+
+func DropHandler(w http.ResponseWriter, r *http.Request) {
+	db.Exec("DROP DATABASE prod")
+}
+```
+
+Good:
+
+```Go
+func main() {
+	db := // ...
+	handlers := Handlers{DB: db}
+	http.HandleFunc("/drop", handlers.DropHandler)
+	// ...
+}
+
+type Handlers struct {
+	DB *sql.DB
+}
+
+func (h *Handlers) DropHandler(w http.ResponseWriter, r *http.Request) {
+	h.DB.Exec("DROP DATABASE prod")
+}
+```
+
+If you really need global variables or constants, e.g., for defining errors or string constants, put them at the top of your file:
+
+```Go
+import "xyz"
+
+const route = "/some-route"
+
+var NotFoundErr = errors.New("not found")
+
+func someFunc() {
+	//...
+}
+
+func someOtherFunc() {
+	// usage of route
+}
+
+func yetAnotherFunc() {
+	// usage of NotFoundErr
+}
+```
+
+## Avoid side-effects
+
+Bad:
+
+```Go
+func init() {
+	someStruct.Load()
+}
+```
+
+Side effects are only okay in special cases (e.g. parsing flags in a cmd). If you find no other way, rethink and refactor.
+
+## Don't over-interface
+
+Bad:
+
+```Go
+type Server interface {
+	Serve() error
+	Some() int
+	Fields() float64
+	That() string
+	Are([]byte) error
+	Not() []string
+	Necessary() error
+}
+```
+
+Good:
+
+```Go
+type Server interface {
+	Serve() error
+}
+
+func run(srv Server) {
+	srv.Serve()
+}
+```
+
+Favour small interfaces and only expect the interfaces you need in your funcs.
+
+## Goimports
+
+https://godoc.org/golang.org/x/tools/cmd/goimports
+
+Command goimports updates your Go import lines, adding missing ones and removing unreferenced ones. In addition to fixing imports, goimports also formats your code in the same style as gofmt so it can be used as a replacement for your editor's gofmt-on-save hook.
```

Comment:
Let's put here the section about import statements ordering and grouping here.

Reply:
updated